roxie/ccd/ccdkey.cpp

```diff
@@ -550,34 +427,79 @@ typedef IArrayOf<InMemoryIndex> InMemoryIndexSet;
 
-class CDirectReaderBase : public CInterface, implements IDirectReader, implements IThorDiskCallback, implements ISimpleReadStream
+class CDirectReaderBase : public CInterface, implements IDirectReader, implements IDirectStreamReader
 {
 protected:
     MemoryBuffer buf;  // Used if translating to hold on to current row;
     CThorContiguousRowBuffer deserializeSource;
     Owned<ISourceRowPrefetcher> prefetcher;
     Linked<const ITranslatorSet> translators;
     const IDynamicTransform *translator = nullptr;
+    const RowFilter &postFilter;
+    RtlDynRow row;
+    bool grouped = false;
+    bool eogPending = false;
+    bool anyThisGroup = false;
 
-public:
-    CDirectReaderBase(const ITranslatorSet *_translators) : translators(_translators) {}
-    virtual const byte *nextRow() override
+    void _nextRow()
     {
         if (prefetcher)
-            prefetcher->readAhead(deserializeSource);
+            prefetcher->readAhead(deserializeSource);  // MORE - is it ever NULL ?
+        if (grouped)
+        {
+            size32_t sizeRead = deserializeSource.queryRowSize();
+            eogPending = (bool) deserializeSource.queryRow()[sizeRead-1];
+        }
         if (translator)
         {
             MemoryBufferBuilder aBuilder(buf, 0);
             translator->translate(aBuilder, deserializeSource.queryRow());
-            return reinterpret_cast<const byte *>(buf.toByteArray());
+            row.setRow(reinterpret_cast<const byte *>(buf.toByteArray()), 0);
         }
         else
-            return deserializeSource.queryRow();
+            row.setRow(deserializeSource.queryRow(), 0);
+    }
+public:
+    CDirectReaderBase(const ITranslatorSet *_translators, const RowFilter &_postFilter, bool _grouped)
+    : translators(_translators), postFilter(_postFilter), row(_translators->queryTargetFormat()), grouped(_grouped)
+    {}
+    virtual bool isKeyed() const override { return false; }
+
+    virtual IDirectStreamReader *queryDirectStreamReader() override
+    {
+        assertex(!translators->isTranslating());
+        return this;
+    };
+    virtual bool eos() = 0;
+
+    virtual void serializeCursorPos(MemoryBuffer &mb) const override
+    {
+        mb.append(tell());
     }
-    virtual bool eog() const override
+
+    virtual const byte *nextRow() override
     {
-        size32_t sizeRead = deserializeSource.queryRowSize();
-        return (bool) deserializeSource.queryRow()[sizeRead-1];
+        while (!eos())
+        {
+            if (eogPending)
+            {
+                eogPending = false;
+                if (anyThisGroup)
+                {
+                    anyThisGroup = false;
+                    return nullptr;
+                }
+            }
+            _nextRow();
+            if (postFilter.matches(row))  // MORE - could filter before translation.
+            {
+                anyThisGroup = true;
```

Comment:
I don't think finishedRow() is called for rows that match - should finishRow() be called at the head of the loop instead

Reply:
The caller is responsible for calling finishRow on the ones that are returned